include/orb/GroundPropagator.h

```diff
@@ -0,0 +1,173 @@
+ * estimator will propagate it to the current time in the background 
+ * while still propagating current as the best estimate. 
+ * Once catching_up is done propagating it replaces current.
+ * 
+ * If another Orbit gets uplinked while catching_up is still being propagated 
+ * in the background, it gets stored in to_catch_up. This ensures too many Orbits getting 
+ * uplinked won't overload the estimator and prevent it from making progress.
+ * If to_catch_up takes fewer grav calls to finish propagating than catching_up 
+ * it replaces catching_up.
+ * Also to_catch_up replaces catching_up if catching_up finishes propagating.
+ */
+class GroundPropagator {
+  private:
+    /** Current Orbit estimate.*/
+    Orbit current;
+
+    /** Orbit catching up to current time.
+     * If catching_up is valid then current is valid.
+     * If catching_up is valid then current.numgravcallsleft()<catching_up.numgravcallsleft().
+     * If catching_up is valid then catching_up is more recently input than current.
+     */
+    Orbit catching_up;
+
+    /** Orbit to catch up to current time once catching_up is done catching up.
+     * If to_catch_up is valid then catching_up is valid.
+     * If to_catch_up is valid then catching_up.numgravcallsleft()<to_catch_up.numgravcallsleft().
+     * If to_catch_up is valid then to_catch_up is more recently input than catching_up.
+     */
+    Orbit to_catch_up;
+
+  public:
+    /**
+     * Construct GroundPropagator.
+     *
+     * grav calls: 0
+     */
+    GroundPropagator(){}
+
+    /** Input the newest Orbit sent from ground.
+     * 
+     * grav calls: 0
+     * @param[in] ground_data: Orbit sent from ground.
+     * @param[in] gps_time_ns: Time to propagate to (ns).
+     * @param[in] earth_rate_ecef: The earth's angular rate in ecef frame ignored for orbits already propagating(rad/s).
+     */
+    void input(const Orbit& ground_data, const uint64_t& gps_time_ns, const lin::Vector3d& earth_rate_ecef){
+        //add to queue
+        if (ground_data.valid()){
+            if(!current.valid()){
+                current= ground_data;
+            } else if(!catching_up.valid()){
+                catching_up= ground_data;
+            } else{
+                to_catch_up= ground_data;
+                //Assuming the new ground_data is closer to the current time 
+                //than previous ground data, over writing this shouldn't be losing any important 
+                //information. 
+                //If that assumion is wrong, then the estimator isn't optimal, but it still works.
+            }
+        }
+        //start propagators
+        current.startpropagating(gps_time_ns,earth_rate_ecef);
+        catching_up.startpropagating(gps_time_ns,earth_rate_ecef);
+        to_catch_up.startpropagating(gps_time_ns,earth_rate_ecef);
+        //sort stuff
+        if (to_catch_up.valid() && to_catch_up.numgravcallsleft()<=catching_up.numgravcallsleft()){
+            catching_up= to_catch_up;
```

Comment:
I don't there are any internal pointers. However, it may be possible to write a more efficient copy assignment operator/constructor if we don't have to copy all of the internal fields into the new object. For example, if we don't need copy assignment while an orbit is propagating and only need to copy over `r`, `v`, and `t` that could save some overhead.

Reply:
Yeah, there are no internal pointers.